src/transaction.c

```diff
@@ -0,0 +1,467 @@
+         bool is_read_entry_locked_by_another =
+            tuple_ts.rts <= tt_txn->commit_ts
+            && lock_table_is_range_locked(txn_kvsb->lock_tbl, rkey, rkey)
+            && tictoc_rw_entry_is_not_in_write_set(tt_txn, r);
+         bool need_to_abort =
+            is_read_entry_written_by_another || is_read_entry_locked_by_another;
+         if (need_to_abort) {
+            platform_mutex_unlock(&txn_kvsb->g_lock);
+            return FALSE;
+         }
+
+         uint32 new_rts            = MAX(tt_txn->commit_ts, tuple_ts.rts);
+         bool   need_to_update_rts = new_rts != tuple_ts.rts;
+         if (need_to_update_rts) {
+            writable_buffer ts;
+            writable_buffer_init(&ts, 0); // FIXME: use a correct heap_id
+            writable_buffer_append(&ts, sizeof(tictoc_timestamp), &new_rts);
+            splinterdb_update(
+               txn_kvsb->kvsb, rkey, writable_buffer_to_slice(&ts));
+            writable_buffer_deinit(&ts);
+         }
+
+         platform_mutex_unlock(&txn_kvsb->g_lock);
+      }
+   }
+
+   return TRUE;
+}
+
+/*
+ * Algorithm 3: Write Phase
+ */
+static void
+tictoc_write(transactional_splinterdb *txn_kvsb, tictoc_transaction *tt_txn)
+{
+   const splinterdb *kvsb = txn_kvsb->kvsb;
+
+   for (uint64 i = 0; i < tt_txn->write_cnt; ++i) {
+      tictoc_rw_entry     *w              = &tt_txn->write_set[i];
+      tictoc_timestamp_set write_entry_ts = get_ts_from_tictoc_rw_entry(w);
+      if (write_entry_ts.wts != tt_txn->commit_ts
+          || write_entry_ts.rts != tt_txn->commit_ts)
+      {
+         write_entry_ts.wts = tt_txn->commit_ts;
+         write_entry_ts.rts = tt_txn->commit_ts;
+
+         tictoc_tuple_header *tuple = writable_buffer_data(&w->tuple);
+         memcpy(&tuple->ts_set, &write_entry_ts, sizeof(tictoc_timestamp_set));
+      }
+
+      slice wkey = writable_buffer_to_slice(&w->key);
+
+      int rc = 0;
+
+      // TODO: merge messages in the write set and write to splinterdb
+      switch (w->op) {
+         case MESSAGE_TYPE_INSERT:
+            rc = splinterdb_insert(
+               kvsb, wkey, writable_buffer_to_slice(&w->tuple));
+            break;
+         case MESSAGE_TYPE_UPDATE:
+            rc = splinterdb_update(
+               kvsb, wkey, writable_buffer_to_slice(&w->tuple));
+            break;
+         case MESSAGE_TYPE_DELETE:
+            rc = splinterdb_delete(kvsb, wkey);
+            break;
+         default:
+            break;
+      }
+
+      platform_assert(rc == 0, "Error from SplinterDB: %d\n", rc);
+
+      writable_buffer_deinit(&w->tuple);
+   }
+}
+
+static int
+tictoc_local_write(transactional_splinterdb *txn_kvsb,
+                   tictoc_transaction       *txn,
+                   tictoc_timestamp_set      ts_set,
+                   slice                     key,
+                   message                   msg)
+{
+   const data_config *cfg =
+      txn_kvsb->tcfg->txn_data_cfg->application_data_config;
+
```

Comment:
The code below overwrites/merges write operations if there already exists the same key in the write set.

This table shows what to do depending on an old operation and new one. If a new operation is Insert/Delete, overwrites the previous operation. Otherwise, merge them.

| New\Old | Insert    | Update    | Delete    |
|---------|-----------|-----------|-----------|
| Insert  | Overwrite | Overwrite | Overwrite |
| Update  | Merge     | Merge     | Invalid   |
| Delete  | Overwrite | Overwrite | Overwrite |